src/main.js

```diff
@@ -0,0 +1,81 @@
+const Apify = require('apify');
+const _ = require('underscore');
+const { getPageRecord, getAndEnqueueLinkUrls } = require('./page-handler');
+const { sendEmailNotification } = require('./notification');
+const { normalizeUrl, setDefaultViewport, getResults, saveResults, getBrokenLinks, saveRecordToDataset, getBaseUrlRequest, urlsOfSameDomain } = require('./tools');
+const { NAVIGATION_TIMEOUT, MAX_REQUEST_RETRIES } = require('./consts');
+
+const { utils: { log } } = Apify;
+
+Apify.main(async () => {
+    const input = await Apify.getValue('INPUT');
+    const { maxConcurrency, maxPages, notificationEmails, saveOnlyBrokenLinks, crawlSubdomains } = input;
+
+    log.info(`Input: ${JSON.stringify(input, null, 2)}`);
+
+    const baseUrl = normalizeUrl(input.baseUrl);
+
+    const requestQueue = await Apify.openRequestQueue();
+    await requestQueue.addRequest(getBaseUrlRequest(baseUrl));
+
+    const records = await Apify.getValue('RECORDS') || [];
+    Apify.events.on('persistState', async () => { await Apify.setValue('RECORDS', records); });
+
+    const { WITH_SUBDOMAINS, WITHOUT_SUBDOMAINS } = MAX_REQUEST_RETRIES;
+    const maxRequestRetries = crawlSubdomains ? WITH_SUBDOMAINS : WITHOUT_SUBDOMAINS;
+
+    const crawler = new Apify.PuppeteerCrawler({
+        requestQueue,
+        maxConcurrency,
+        maxRequestsPerCrawl: maxPages,
+        maxRequestRetries,
+        browserPoolOptions: {
+            preLaunchHooks: [setDefaultViewport]
+        },
+        navigationTimeoutSecs: NAVIGATION_TIMEOUT,
+        handlePageFunction: async (context) => {
+            const { request: { url } } = context;
+            log.info(`Crawling page...`, { url });
+
+            const record = await getPageRecord(context, crawlSubdomains);
+
+            // If we're on the base website or we're allowed to crawl current subdomain, find links to new pages and enqueue them.
+            const crawlCurrentSubdomain = crawlSubdomains && urlsOfSameDomain(baseUrl, url);
+            if (record.isBaseWebsite || crawlCurrentSubdomain) {
+                record.linkUrls = await getAndEnqueueLinkUrls(context);
+            }
+
+            await saveRecordToDataset(record, saveOnlyBrokenLinks);
+            records.push(record);
+        },
+
+        // This function is called if the page processing failed more than maxRequestRetries+1 times.
+        handleFailedRequestFunction: async ({ request }) => {
+            const url = normalizeUrl(request.url);
+            log.info(`Page failed ${request.retryCount + 1} times, giving up: ${url}`);
```

Comment:
Better to make a warning out of important logs, which are not errors - `log.warning`